File: apps/portal/src/lib/api/schemas/blog.ts
import { z } from 'zod';

export const BLOG_LOCALES = ['en', 'nl'] as const;
export const BLOG_STATUSES = ['draft', 'published'] as const;

export const BlogPostReadSchema = z.object({
  id: z.string(),
  slug: z.string(),
  locale: z.string(),
  title: z.string(),
  description: z.string(),
  author: z.string(),
  tags: z.array(z.string()),
  published_at: z.string(),
  cover_image_url: z.string(),
  cover_image_key: z.string(),
  content_key: z.string(),
  content: z.union([z.string(), z.null()]).optional(),
  status: z.string(),
  created_by_user_id: z.union([z.string(), z.null()]),
  created_at: z.string(),
  updated_at: z.string(),
});

export type BlogPostRead = z.infer<typeof BlogPostReadSchema>;

export const BlogPostListSchema = z.array(BlogPostReadSchema);

export const BlogPostUpdateInputSchema = z.object({
  title: z.string().min(1).max(255).optional(),
  description: z.string().min(1).max(2000).optional(),
  slug: z.string().min(1).max(160).optional(),
  locale: z.enum(BLOG_LOCALES).optional(),
  author: z.string().max(120).optional(),
  tags: z.array(z.string()).optional(),
  published_at: z.string().optional(),
  status: z.enum(BLOG_STATUSES).optional(),
  content: z.string().optional(),
});

export type BlogPostUpdateInput = z.infer<typeof BlogPostUpdateInputSchema>;

// Body for `POST /admin/blog/posts` — multipart payload composed in
// `lib/api/blog.ts`. The image and content live alongside structured
// metadata; the route packs them into a `FormData` before sending.
export type BlogPostCreateInput = {
  slug: string;
  locale: 'en' | 'nl';
  title: string;
  description: string;
  author?: string;
  tags: string[];
  // ISO-8601 with timezone — typically `new Date().toISOString()`. The API
  // rejects naive timestamps with `BLOG_PUBLISHED_AT_INVALID`.
  published_at: string;
  content: string;
  status: 'draft' | 'published';
  cover: File;
};

// `POST /admin/blog/posts/bilingual` returns one row per locale.
export const BlogPostBilingualResponseSchema = z.object({
  en: BlogPostReadSchema,
  nl: BlogPostReadSchema,
});

export type BlogPostBilingualResponse = z.infer<
  typeof BlogPostBilingualResponseSchema
>;

// Body for `POST /admin/blog/posts/bilingual`. The shared fields (including
// a single cover image + single description) are sent once; the EN and NL
// halves carry only their own title + content. Backend stores the same
// cover_image_key + description on both rows.
export type BlogPostBilingualCreateInput = {
  slug: string;
  author?: string;
  tags: string[];
  published_at: string;
  status: 'draft' | 'published';
  description: string;
  cover: File;
  en: {
    title: string;
    content: string;
  };
  nl: {
    title: string;
    content: string;
  };
};

File: apps/api/src/bimstitch_api/schemas/blog.py
"""Pydantic schemas for the blog admin + public endpoints."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

BLOG_LOCALES = {"en", "nl"}
BLOG_STATUSES = {"draft", "published"}


class BlogPostRead(BaseModel):
    """Full read shape returned to the admin UI.

    `cover_image_url` and `content_url` are freshly-presigned URLs generated
    at read time — clients must not cache them past their TTL (default 15 min).
    `content` is the inlined Markdown body for the detail view; list endpoints
    omit it to keep the payload small.
    """

    id: UUID
    slug: str
    locale: str
    title: str
    description: str
    author: str
    tags: list[str]
    published_at: datetime
    cover_image_url: str
    cover_image_key: str
    content_key: str
    content: str | None = None
    status: str
    created_by_user_id: UUID | None
    created_at: datetime
    updated_at: datetime


class BlogPostPublicRead(BaseModel):
    """Slimmer read shape for the public marketing site.

    Strips internal fields (status, created_by, S3 keys) — the web app only
    needs what it renders. Cover URL is presigned and short-lived; web app
    should re-fetch on cache miss rather than persist.
    """

    slug: str
    locale: str
    title: str
    description: str
    author: str
    tags: list[str]
    published_at: datetime
    cover_image_url: str
    content: str | None = None
    reading_time_minutes: int


class BlogPostBilingualResponse(BaseModel):
    """Pair of newly-created posts returned by the bilingual create endpoint.

    The endpoint atomically writes one English row and one Dutch row sharing
    the same slug; this response surfaces both so the admin UI can immediately
    render the canonical (slug, locale) pair without a follow-up GET.
    """

    en: BlogPostRead
    nl: BlogPostRead


class BlogPostUpdate(BaseModel):
    """Optional fields for PATCH /admin/blog/posts/{id}.

    `cover_image` and the markdown body are handled as multipart parts in
    the route signature, not in this body — this schema is JSON-only.
    """

    title: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = Field(default=None, min_length=1, max_length=2000)
    slug: str | None = Field(default=None, min_length=1, max_length=160)
    locale: str | None = None
    author: str | None = Field(default=None, max_length=120)
    tags: list[str] | None = None
    published_at: datetime | None = None
    status: str | None = None
    content: str | None = None
